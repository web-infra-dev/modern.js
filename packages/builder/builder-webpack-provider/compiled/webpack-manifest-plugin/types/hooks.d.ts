import { Compiler, Module, Compilation, LoaderContext } from 'webpack';
import { EmitCountMap, InternalOptions } from './';
interface BeforeRunHookArgs {
    emitCountMap: EmitCountMap;
    manifestFileName: string;
}
interface EmitHookArgs {
    compiler: Compiler;
    emitCountMap: EmitCountMap;
    manifestAssetId: string;
    manifestFileName: string;
    moduleAssets: Record<any, any>;
    options: InternalOptions;
}
declare const getCompilerHooks: (compiler: Compiler) => any;
declare const beforeRunHook: ({ emitCountMap, manifestFileName }: BeforeRunHookArgs, _: Compiler, callback: Function) => void;
declare const emitHook: ({ compiler, emitCountMap, manifestAssetId, manifestFileName, moduleAssets, options }: EmitHookArgs, compilation: Compilation) => void;
interface LegacyModule extends Module {
    userRequest?: any;
}
declare const normalModuleLoaderHook: ({ moduleAssets }: {
    moduleAssets: Record<any, any>;
}, loaderContext: LoaderContext<any>, module: LegacyModule) => void;
export { beforeRunHook, emitHook, getCompilerHooks, normalModuleLoaderHook };
