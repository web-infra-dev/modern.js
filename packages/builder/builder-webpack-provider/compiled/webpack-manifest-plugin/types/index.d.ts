import { Compiler, WebpackPluginInstance } from 'webpack';
import { FileDescriptor } from './helpers';
import { getCompilerHooks } from './hooks';
export declare type Manifest = Record<string, any>;
export interface InternalOptions {
    [key: string]: any;
    assetHookStage: number;
    basePath: string;
    fileName: string;
    filter: (file: FileDescriptor) => Boolean;
    generate: (seed: Record<any, any>, files: FileDescriptor[], entries: Record<string, string[]>) => Manifest;
    map: (file: FileDescriptor) => FileDescriptor;
    publicPath: string;
    removeKeyHash: RegExp | false;
    seed: Record<any, any>;
    serialize: (manifest: Manifest) => string;
    sort: (fileA: FileDescriptor, fileB: FileDescriptor) => Number;
    transformExtensions: RegExp;
    useEntryKeys: Boolean;
    useLegacyEmit: Boolean;
    writeToFileEmit: Boolean;
}
export declare type ManifestPluginOptions = Partial<InternalOptions>;
export declare type EmitCountMap = Map<any, any>;
declare class WebpackManifestPlugin implements WebpackPluginInstance {
    private options;
    constructor(opts: ManifestPluginOptions);
    apply(compiler: Compiler): void;
}
export { getCompilerHooks, WebpackManifestPlugin };
