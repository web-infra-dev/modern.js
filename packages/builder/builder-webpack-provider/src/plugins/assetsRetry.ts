import { BuilderPlugin } from '../types';
import { isHtmlDisabled } from './html';

export function builderPluginAssetsRetry(): BuilderPlugin {
  return {
    name: 'builder-plugin-assets-retry',
    setup(api) {
      api.modifyWebpackChain(async (chain, { CHAIN_ID, target }) => {
        const config = api.getNormalizedConfig();

        if (!config.output.assetsRetry || isHtmlDisabled(config, target)) {
          return;
        }

        const { AssetsRetryPlugin } = await import(
          '../webpackPlugins/AssetsRetryPlugin'
        );
        chain
          .plugin(CHAIN_ID.PLUGIN.ASSETS_RETRY)
          .use(AssetsRetryPlugin, [config.output.assetsRetry]);
      });
    },
  };
}
