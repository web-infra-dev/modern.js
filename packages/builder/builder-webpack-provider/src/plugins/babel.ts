import {
  getBabelConfig,
  createBabelChain,
  type BabelOptions,
} from '@modern-js/babel-preset-app';
import {
  JS_REGEX,
  TS_REGEX,
  mergeRegex,
  createVirtualModule,
  getBrowserslistWithDefault,
  type BuilderContext,
} from '@modern-js/builder-shared';

import type { WebpackChain, BuilderPlugin, NormalizedConfig } from '../types';

const enableCoreJsEntry = (config: NormalizedConfig, isServer: boolean) =>
  config.output.polyfill === 'entry' && !isServer;

export const getUseBuiltIns = (config: NormalizedConfig) => {
  const { polyfill } = config.output;
  if (polyfill === 'ua' || polyfill === 'off') {
    return false;
  }
  return polyfill;
};

export function applyScriptCondition({
  rule,
  config,
  context,
  includes,
  excludes,
}: {
  rule: WebpackChain.Rule;
  config: NormalizedConfig;
  context: BuilderContext;
  includes: (string | RegExp)[];
  excludes: (string | RegExp)[];
}) {
  // compile all folders in app directory, exclude node_modules
  rule.include.add({
    and: [context.rootPath, { not: /node_modules/ }],
  });

  [...includes, ...(config.source.include || [])].forEach(condition => {
    rule.include.add(condition);
  });

  [...excludes, ...(config.source.exclude || [])].forEach(condition => {
    rule.exclude.add(condition);
  });
}

export const builderPluginBabel = (): BuilderPlugin => ({
  name: 'builder-plugin-babel',
  setup(api) {
    api.modifyWebpackChain(
      async (
        chain,
        { CHAIN_ID, getCompiledPath, target, isProd, isServer },
      ) => {
        const { applyOptionsChain, isUseSSRBundle } = await import(
          '@modern-js/utils'
        );

        const config = api.getNormalizedConfig();
        const browserslist = await getBrowserslistWithDefault(
          api.context.rootPath,
          config,
          target,
        );

        const getBabelOptions = (
          appDirectory: string,
          config: NormalizedConfig,
        ) => {
          // 1. Get styled-components options
          const styledComponentsOptions = applyOptionsChain(
            {
              // "pure" is used to improve dead code elimination in production.
              // we don't need to enable it in development because it will slow down the build process.
              pure: isProd,
              displayName: true,
              ssr: isUseSSRBundle(config),
              transpileTemplateLiterals: true,
            },
            config.tools.styledComponents,
          );

          // 2. Create babel util function about include/exclude
          const includes: Array<string | RegExp> = [];
          const excludes: Array<string | RegExp> = [];

          const babelUtils = {
            addIncludes(items: string | RegExp | Array<string | RegExp>) {
              if (Array.isArray(items)) {
                includes.push(...items);
              } else {
                includes.push(items);
              }
            },
            addExcludes(items: string | RegExp | Array<string | RegExp>) {
              if (Array.isArray(items)) {
                excludes.push(...items);
              } else {
                excludes.push(items);
              }
            },
          };

          // 3. Compute final babel config by @modern-js/babel-preset-app
          const babelOptions: BabelOptions = {
            babelrc: false,
            configFile: false,
            compact: isProd,
            ...getBabelConfig({
              target: isServer ? 'server' : 'client',
              appDirectory,
              useLegacyDecorators: !config.output.enableLatestDecorators,
              useBuiltIns: isServer ? false : getUseBuiltIns(config),
              chain: createBabelChain(),
              styledComponents: styledComponentsOptions,
              userBabelConfig: config.tools.babel,
              userBabelConfigUtils: babelUtils,
              overrideBrowserslist: browserslist,
            }),
          };

          if (config.output.charset === 'utf8') {
            babelOptions.generatorOpts = {
              jsescOption: { minimal: true },
            };
          }

          return {
            babelOptions,
            includes,
            excludes,
          };
        };

        const { rootPath } = api.context;
        const { babelOptions, includes, excludes } = getBabelOptions(
          rootPath,
          config,
        );
        const useTsLoader = Boolean(config.tools.tsLoader);
        const rule = chain.module.rule(CHAIN_ID.RULE.JS);

        applyScriptCondition({
          rule,
          config,
          context: api.context,
          includes,
          excludes,
        });

        rule
          .test(useTsLoader ? JS_REGEX : mergeRegex(JS_REGEX, TS_REGEX))
          .use(CHAIN_ID.USE.BABEL)
          .loader(getCompiledPath('babel-loader'))
          .options(babelOptions);

        /**
         * If a script is imported with data URI, it can be compiled by babel too.
         * This is used by some frameworks to create virtual entry.
         * https://webpack.js.org/api/module-methods/#import
         * @example: import x from 'data:text/javascript,export default 1;';
         */
        if (config.source.compileJsDataURI) {
          chain.module
            .rule(CHAIN_ID.RULE.JS_DATA_URI)
            .mimetype({
              or: ['text/javascript', 'application/javascript'],
            })
            .use(CHAIN_ID.USE.BABEL)
            .loader(getCompiledPath('babel-loader'))
            .options(babelOptions);
        }

        addCoreJsEntry({ chain, config, isServer });
      },
    );
  },
});

/** Add core-js-entry to every entries. */
export function addCoreJsEntry({
  chain,
  config,
  isServer,
}: {
  chain: WebpackChain;
  config: NormalizedConfig;
  isServer: boolean;
}) {
  if (enableCoreJsEntry(config, isServer)) {
    const entryPoints = Object.keys(chain.entryPoints.entries() || {});
    const coreJsEntry = createVirtualModule('import "core-js";');

    for (const name of entryPoints) {
      chain.entry(name).prepend(coreJsEntry);
    }
  }
}
