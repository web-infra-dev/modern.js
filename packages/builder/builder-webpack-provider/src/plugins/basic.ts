import type { BuilderPlugin } from '../types';

/**
 * Provide some basic configs of webpack
 */
export const builderPluginBasic = (): BuilderPlugin => ({
  name: 'builder-plugin-basic',

  setup(api) {
    api.modifyWebpackChain(async (chain, { isProd }) => {
      // The base directory for resolving entry points and loaders from the configuration.
      chain.context(api.context.rootPath);

      chain.mode(isProd ? 'production' : 'development');

      chain.merge({
        infrastructureLogging: {
          // Using `error` level to avoid `webpack.cache.PackFileCacheStrategy` logs
          level: 'error',
        },
      });

      // if the chunk size exceeds 1MiB, we will throw a warning
      chain.performance.maxAssetSize(1024 * 1024);
      chain.performance.maxEntrypointSize(1024 * 1024);

      // This will be futureDefaults in webpack 6
      chain.module.parser.merge({
        javascript: {
          exportsPresence: 'error',
        },
      });
    });
  },
});
