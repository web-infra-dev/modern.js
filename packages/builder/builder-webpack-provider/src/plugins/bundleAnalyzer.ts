import { BuilderPlugin } from '../types';

export function builderPluginBundleAnalyzer(): BuilderPlugin {
  return {
    name: 'builder-plugin-bundle-analyzer',
    setup(api) {
      api.modifyWebpackChain(
        async (chain, { CHAIN_ID, getCompiledPath, target }) => {
          const config = api.getNormalizedConfig();
          // There are two ways to open the bundle analyzer:
          // 1. Set environment variable `BUNDLE_ANALYZE`
          // 2. Set performance.bundleAnalyze config
          if (
            !process.env.BUNDLE_ANALYZE &&
            !config.performance.bundleAnalyze
          ) {
            return;
          }
          const { default: BundleAnalyzer } = await import(
            getCompiledPath('webpack-bundle-analyzer')
          );
          chain
            .plugin(CHAIN_ID.PLUGIN.BUNDLE_ANALYZER)
            .use(BundleAnalyzer.BundleAnalyzerPlugin, [
              {
                analyzerMode: 'static',
                openAnalyzer: false,
                reportFilename: `report-${target}.html`,
                ...(config.performance.bundleAnalyze || {}),
              },
            ]);
        },
      );
    },
  };
}
