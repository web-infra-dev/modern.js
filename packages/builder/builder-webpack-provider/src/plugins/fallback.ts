import { join } from 'path';
import {
  JS_REGEX,
  TS_REGEX,
  getDistPath,
  getFilename,
} from '@modern-js/builder-shared';
import type { BuilderPlugin } from '../types';
import type { RuleSetRule } from 'webpack';

export const PluginFallback = (): BuilderPlugin => ({
  name: 'builder-plugin-fallback',

  setup(api) {
    api.modifyWebpackChain((chain, { isProd }) => {
      const builderConfig = api.getNormalizedConfig();

      if (builderConfig.output.enableAssetFallback) {
        const distDir = getDistPath(builderConfig.output, 'media');
        const filename = getFilename(builderConfig.output, 'media', isProd);

        chain.output.merge({
          assetModuleFilename: join(distDir, filename),
        });
      }
    });

    api.modifyWebpackConfig(config => {
      const builderConfig = api.getNormalizedConfig();

      if (!builderConfig.output.enableAssetFallback || !config.module) {
        return;
      }

      const { rules = [] } = config.module;

      const innerRules: Array<RuleSetRule> = [];
      const outerRules: Array<RuleSetRule | '...'> = [];

      for (const rule of rules) {
        if (
          // "..." refers to the webpack defaults
          rule === '...' ||
          // this is a special case, put the mjs fullySpecified rule in the outside
          (rule.resolve && !rule.mimetype)
        ) {
          outerRules.push(rule);
        } else if (
          rule.oneOf &&
          !(
            rule.test ||
            rule.exclude ||
            rule.resource ||
            rule.issuer ||
            rule.mimetype
          )
        ) {
          rule.oneOf.forEach(r => innerRules.push(r));
        } else {
          innerRules.push(rule);
        }
      }

      const fileLoader = {
        exclude: [
          JS_REGEX,
          TS_REGEX,
          // exclude `html` and `json`, they get processed by webpack internal loaders.
          /\.html$/,
          /\.json$/,
        ],
        type: 'asset/resource',
      };

      config.module.rules = [
        ...outerRules,
        { oneOf: [...innerRules, fileLoader] },
      ];
    });
  },
});
