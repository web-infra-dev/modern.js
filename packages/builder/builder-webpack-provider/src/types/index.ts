/* eslint-disable @typescript-eslint/no-restricted-imports */
export * from './hooks';
export * from './config';
export * from './plugin';
export * from './context';
export * from './thirdParty';
/* eslint-enable @typescript-eslint/no-restricted-imports */
