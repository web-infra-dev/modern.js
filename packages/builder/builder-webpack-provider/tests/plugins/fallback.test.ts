import { expect, describe, it } from 'vitest';
import { builderPluginFallback } from '@/plugins/fallback';
import { BuilderPlugin } from '@/types';
import { createStubBuilder } from '@/stub';

describe('plugins/fallback', () => {
  const testPlugin: BuilderPlugin = {
    name: 'test-plugin',
    setup(api) {
      api.modifyWebpackChain(chain => {
        chain.module
          .rule('mjs')
          .test(/\.m?js/)
          .resolve.set('fullySpecified', false);

        chain.module
          .rule('foo')
          .oneOf('foo')
          .test(/foo/)
          .use('foo')
          .loader('foo');

        chain.module
          .rule('data-uri')
          .resolve.set('fullySpecified', false)
          .end()
          .mimetype('text/javascript')
          .use('data-uri')
          .loader('data-uri');

        chain.module.rule('bar').test(/bar/).use('bar').loader('bar');
      });
    },
  };

  it('should convert fallback rule correctly', async () => {
    const builder = await createStubBuilder({
      plugins: [testPlugin, builderPluginFallback()],
      builderConfig: {
        output: {
          enableAssetFallback: true,
        },
      },
    });
    const config = await builder.unwrapWebpackConfig();

    expect(config).toMatchSnapshot();
  });

  it('should not convert fallback rule when output.enableAssetFallback is not enabled', async () => {
    const builder = await createStubBuilder({
      plugins: [testPlugin, builderPluginFallback()],
    });
    const config = await builder.unwrapWebpackConfig();

    expect(config).toMatchSnapshot();
  });
});
