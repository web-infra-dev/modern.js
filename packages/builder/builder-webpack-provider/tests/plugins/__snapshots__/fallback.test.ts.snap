// Vitest Snapshot v1

exports[`plugins/fallback > should convert fallback rule correctly 1`] = `
{
  "module": {
    "rules": [
      {
        "resolve": {
          "fullySpecified": false,
        },
        "test": /\\\\\\.m\\?js/,
      },
      {
        "oneOf": [
          {
            "test": /foo/,
            "use": [
              {
                "loader": "foo",
              },
            ],
          },
          {
            "test": /bar/,
            "use": [
              {
                "loader": "bar",
              },
            ],
          },
          {
            "exclude": [
              /\\\\\\.\\(js\\|mjs\\|cjs\\|jsx\\)\\$/,
              /\\\\\\.\\(ts\\|mts\\|cts\\|tsx\\)\\$/,
              /\\\\\\.html\\$/,
              /\\\\\\.json\\$/,
            ],
            "type": "asset/resource",
          },
        ],
      },
    ],
  },
}
`;

exports[`plugins/fallback > should not convert fallback rule when output.enableAssetFallback is not enabled 1`] = `
{
  "module": {
    "rules": [
      {
        "resolve": {
          "fullySpecified": false,
        },
        "test": /\\\\\\.m\\?js/,
      },
      {
        "oneOf": [
          {
            "test": /foo/,
            "use": [
              {
                "loader": "foo",
              },
            ],
          },
        ],
      },
      {
        "test": /bar/,
        "use": [
          {
            "loader": "bar",
          },
        ],
      },
    ],
  },
}
`;
